### PR TITLE
Rfe eap7 583

### PR DESCRIPTION
**Pull request by jstefl**

### CODE REVIEW

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/DeploymentPage.java [hunk 1]
```diff
@@ -16,11 +20,36 @@
...
+            return "";
+        }
+
+        WebElement content = getContentRoot().findElement(new ByJQuery(selector.toString()));
+        return content.getText();
+    }
+
+    public List<String> getDeploymentBrowsedContentItems() {
+        By selector = ByJQuery.selector("table.browse-content tr");
+        List<String> items = new ArrayList();
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/DeploymentPage.java]: Please add `String` type args: `new ArrayList<String>()`

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/DomainDeploymentPage.java [hunk 2]
```diff
@@ -102,4 +105,16 @@ public Column navigateToDeploymentColumnInServerGroup(String groupName) {
         Console.withBrowser(browser).waitUntilLoaded();
         return column;
     }
+
+    public List<String> getDeploymentBrowsedContentItems() {
+        By selector = ByJQuery.selector("table.browse-content tr");
+        List<String> items = new ArrayList();
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/DomainDeploymentPage.java]: Please add `String` type args: `new ArrayList<String>()`

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java [hunk 3]
```diff
@@ -161,4 +302,194 @@ public void removeDeployment() throws Exception {
...
 
+    @Test
+    public void unmanagedDeployment() throws Exception {
+        ops.viewDeployment(browser, navigation, DEPLOYMENT_MED_10);
+
+        String content = page.getDeploymentContentTableRow("Managed");
+        assertTrue(content.contains("false"));
+
+        content = page.getDeploymentContentTableRow("Managed");
+        assertTrue(content.contains("false"));
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java]: What's the difference between last two assertions?

File: common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java [hunk 4]
```diff
@@ -224,6 +225,17 @@ public NotificationCenterFragment getNotificationArea() {
         return notification;
     }
 
+    /**
+     * @return returns area display message as popup element in center of page, Area is closed after few seconds automatically usually.
+     */
+    public AlertFragment getAlertArea() {
+        WebElement element = browser.findElement(By.cssSelector(AlertFragment.ROOT_SELECTOR));
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java]: I suggest to add `Graphene.waitGui()` until this element is visible.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java [hunk 5]
```diff
@@ -161,4 +302,194 @@ public void removeDeployment() throws Exception {
...
+        Row row = navigation.step(FinderNames.DEPLOYMENT, DEPLOYMENT_MED_1).selectRow();
+        Console.withBrowser(browser).dismissReloadRequiredWindowIfPresent();
+
+        row.invoke(FinderNames.EXPLODE);
+
+        // TODO - remove whole coment before merge
+        // * Can problems with deplayed popup occurred?
+        //   If so, how can I prevent them? Is active waiting possible or smth else?
+        // * I am aware of class DeploymentOperations for extraction of verification out of test
+        //   I don't want to use it becase my verification is relatively simple, self explanatory.
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java]: Don't forget to remove this comment.
Regarding the question see my comment on `BasePage.getAlertArea()`

File: common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java [hunk 6]
```diff
@@ -45,10 +45,28 @@ public String getIncludedFromProfile() {
         return element.getText().substring(22).trim();
     }
 
-    public List<String> getDisplayedUnorderedList() {
-        By selector = ByJQuery.selector("ul li");
-        Graphene.waitAjax().until().element(root, selector).is().visible();
+    public List<String> getDisplayedUnorderedList(String waitUntilValue) {
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java:56]: Add javadoc. Please describe returned list and especially `waitUntilValue` parameter.

File: common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java [hunk 7]
```diff
@@ -45,10 +45,28 @@ public String getIncludedFromProfile() {
...
+        By selector = ByJQuery.selector(selectorString);
+        By selectorWait = ByJQuery.selector(selectorWaitBuffer.toString());
+
+        Graphene.waitModel(browser).until().element(selectorWait).is().visible();
+
         return root.findElements(selector).stream().filter(element -> element.isDisplayed()).map(element -> element.getText()).collect(toList());
     }
 
+    /**
+     * @return Text place above unordered list usually.
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java]: Please make this sentence more clear. Something like 'Text usually placed above unordered list.' - if I understand your intention?

File: common/src/main/java/org/jboss/hal/testsuite/fragment/AlertFragment.java [hunk 8]
```diff
@@ -0,0 +1,10 @@
+package org.jboss.hal.testsuite.fragment;
+
+
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/fragment/AlertFragment.java:3]: Please add javadoc.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 9]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
         verifyDeployementEnabledInServerGroup(serverGroup, deploymentName, false);
     }
+
+    public void verifyDeploymentContentDefault(List<String> itemsInDeploment) {
+        assertTrue(itemInListExistsLowerCase(itemsInDeploment, "meta-inf/manifest.mf | 127 b"));
+        assertTrue(itemInListExistsLowerCase(itemsInDeploment, "index.jsp | 361 b"));
+        assertTrue(itemInListExistsLowerCase(itemsInDeploment, "page.html | 68 b"));
+    }
+
+    public void enabledDisabledDeployment(WebDriver browser, FinderNavigation navigation, String deploymentName) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Please 
- move this method to a suitable UI abstraction e.g. to `DeploymentPage`.
- rename this method to `enableDisabledDeployment` (without the confusing expletive 'd')
- add javadoc describing at least the `navigation` parameter

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 10]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        Row row = navigation.step(FinderNames.DEPLOYMENT, deploymentName).selectRow();
+        Console.withBrowser(browser).dismissReloadRequiredWindowIfPresent();
+        row.invoke(FinderNames.ENABLE);
+        Console.withBrowser(browser).openedWindow(ConfirmationWindow.class).confirm();
+    }
+
+    public void disableEnabledDeployment(WebDriver browser, FinderNavigation navigation, String deploymentName) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Please 
- move this method to a suitable UI abstraction e.g. to `DeploymentPage`.
- add javadoc describing at least the `navigation` parameter

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 11]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        Row row = navigation.step(FinderNames.DEPLOYMENT, deploymentName).selectRow();
+        Console.withBrowser(browser).dismissReloadRequiredWindowIfPresent();
+        row.invoke(FinderNames.DISABLE);
+        Console.withBrowser(browser).openedWindow(ConfirmationWindow.class).confirm();
+    }
+
+    public void viewDeployment(WebDriver browser, FinderNavigation navigation, String deploymentName) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Please 
- move this method to a suitable UI abstraction e.g. to `DeploymentPage`.
- add javadoc describing at least the `navigation` parameter

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 12]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+
+    public boolean deplomentEnabledTimestampExists(List<String> previewListItems) {
+        return itemInListExistsLowerCase(previewListItems, "last enabled at");
+    }
+
+    public boolean deplomentDisabledTimestampExists(List<String> previewListItems) {
+        return itemInListExistsLowerCase(previewListItems, "last disabled at");
+    }
+
+    protected boolean itemInListExistsLowerCase(List<String> lines, String item) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Why not rather package private visibility? Protected visibility seems to recommend subclassing of `DeploymentsOperations`. Was it your intention? Why?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 13]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        Console.withBrowser(browser).openedWindow(ConfirmationWindow.class).confirm();
+    }
+
+    public void viewDeployment(WebDriver browser, FinderNavigation navigation, String deploymentName) {
+        Row row = navigation.step(FinderNames.DEPLOYMENT, deploymentName).selectRow();
+        Console.withBrowser(browser).dismissReloadRequiredWindowIfPresent();
+        row.invoke(FinderNames.VIEW);
+    }
+
+    protected boolean deploymentIsEnabled(String introText) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Either use this method or remove it.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 14]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        return introText.toLowerCase().contains("deployment is enabled");
+    }
+
+    protected boolean deploymentIsManaged(List<String> previewLines) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Either use this method or remove it.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 15]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        return itemInListExistsLowerCase(previewLines, "managed: yes");
+    }
+
+    protected boolean deploymentIsArchive(List<String> previewLines) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Either use this method or remove it.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 16]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        return itemInListExistsLowerCase(previewLines, "archive: yes");
+    }
+
+    protected boolean deploymentNeverEnabled(List<String> previewListItems) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Why not rather package private visibility? Protected visibility seems to recommend subclassing of DeploymentsOperations. Was it your intention? Why?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java [hunk 17]
```diff
@@ -94,4 +103,69 @@ public void verifyIsDeploymentEnabledInServerGroup(String serverGroup, String de
...
+        return itemInListExistsLowerCase(previewListItems, "the deployment was never enabled");
+    }
+
+    protected boolean deploymentNeverDisabled(List<String> previewListItems) {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Why not rather package private visibility? Protected visibility seems to recommend subclassing of DeploymentsOperations. Was it your intention? Why?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DomainManagedDeploymentsTestCase.java [hunk 18]
```diff
@@ -61,18 +66,30 @@
...
+
+    private static final String DEPLOYMENT_MED_2 = "n-med_dom_2" + RandomStringUtils.randomAlphanumeric(5) + ".war";
+    private static final String DEPLOYMENT_MED_2_RUNTIMME_NAME = "rn-med_dom_2" + RandomStringUtils.randomAlphanumeric(5) + ".war";
+
 
     private static OnlineManagementClient client = ManagementClientProvider.createOnlineManagementClient();
     private static Operations operations = new Operations(client);
     private static Administration administration = new Administration(client);
     private static DeploymentsOperations ops = new DeploymentsOperations(client);
+    private FinderNavigation navigation;
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DomainManagedDeploymentsTestCase.java]: Seems this field is never used. Please remove.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DomainManagedDeploymentsTestCase.java [hunk 19]
```diff
@@ -61,18 +66,30 @@
...
 
     @Drone
     WebDriver browser;
 
     @Page
     DomainDeploymentPage page;
 
+    @Before
+    public void before() {
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DomainManagedDeploymentsTestCase.java]: Unnecessary.  Please remove this method.

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java [hunk 20]
```diff
@@ -54,6 +64,45 @@
     private static final String DEPLOYMENT_ENABLED_NAME = "n-enabled_" + RandomStringUtils.randomAlphanumeric(5) + ".war";
     private static final String DEPLOYMENT_ENABLED_RUNTIME_NAME = "rn-enabled_" + RandomStringUtils.randomAlphanumeric(5) + ".war";
 
+    private static final String DEPLOYMENT_MED_1 = "n-med_1" + RandomStringUtils.randomAlphanumeric(5) + ".war";
```
Comment (zebrik) [inline basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/ManagedDeploymentsTestCase.java]: Is it possible to have more descriptive deployment names indicating if they are enabled or managed? WDYT?

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/DomainDeploymentPage.java [hunk 21]
```diff
@@ -102,4 +105,16 @@ public Column navigateToDeploymentColumnInServerGroup(String groupName) {
         Console.withBrowser(browser).waitUntilLoaded();
         return column;
     }
+
+    public List<String> getDeploymentBrowsedContentItems() {
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/DomainDeploymentPage.java:109]: This code duplication between `DeploymentPage` and `DomainDeploymentPage` will be fixed in another PR. You don't need to fix this.

**Review by zebrik — APPROVED** [review]
I have some remarks to this PR. @jstefl can you please look at them?

File: common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java (hunk 4 above)
Comment (jstefl) [reply common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java]: You meas just add `Graphene.waitGui();` to line 233 ?

File: basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java (hunk 12 above)
Comment (jstefl) [reply basic/src/test/java/org/jboss/hal/testsuite/test/runtime/deployments/DeploymentsOperations.java]: Hm, I am not sure why I let it as protected. I should make it public to ne consistet with other methods in class.

**Comment by jstefl** [comment]
Thank you for review, the changes are incorporated into new commit.

File: common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java (hunk 4 above)
Comment (zebrik) [reply common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java]: I meant something like `Graphene.waitGui().until().element(By.cssSelector(AlertFragment.ROOT_SELECTOR)).is().visible();`

**Comment by jstefl** [comment]
Thank you, updated.

File: common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java [hunk 22]
```diff
@@ -224,6 +225,19 @@ public NotificationCenterFragment getNotificationArea() {
         return notification;
     }
 
+    /**
+     * @return returns area display message as popup element in center of page, Area is closed after few seconds automatically usually.
+     */
+    public AlertFragment getAlertArea() {
+        Graphene.waitGui().until().element(By.cssSelector(AlertFragment.ROOT_SELECTOR)).is().visible();
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/BasePage.java]: Can you define selector like `By alertSelector = By.cssSelector(AlertFragment.ROOT_SELECTOR);` and then use it twice instead of code duplication? It would be more readable and maintanable IMO.

File: common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java [hunk 23]
```diff
@@ -45,10 +45,36 @@ public String getIncludedFromProfile() {
...
     }
 
-    public List<String> getDisplayedUnorderedList() {
-        By selector = ByJQuery.selector("ul li");
-        Graphene.waitAjax().until().element(root, selector).is().visible();
+    /**
+     * Extract unordered html-list from page.
+     *
+     * @param waitUntilValue  method will finishes successfully only when specified value occurs in list.
+     *                        It waits for global timeout at worst.
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java]: I suggest to change `for global timeout` to `for global waitModelInterval timeout` since we have more predefined timeouts.

File: common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java [hunk 24]
```diff
@@ -45,10 +45,36 @@ public String getIncludedFromProfile() {
...
+        By selector = ByJQuery.selector(selectorString);
+        By selectorWait = ByJQuery.selector(selectorWaitBuffer.toString());
+
+        Graphene.waitModel(browser).until().element(selectorWait).is().visible();
+
         return root.findElements(selector).stream().filter(element -> element.isDisplayed()).map(element -> element.getText()).collect(toList());
     }
 
+    /**
+     * @return Text placed above unordered list usually.
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/finder/PreviewFragment.java]: Please make this sentence more clear. Something like 'Text usually placed above unordered list.' - if I understand your intention correctly?

File: common/src/main/java/org/jboss/hal/testsuite/page/runtime/DeploymentPage.java [hunk 25]
```diff
@@ -16,11 +25,59 @@
...
+
+        List<WebElement> webElements = getContentRoot().findElements(selector);
+        for (WebElement webElement : webElements ) {
+            items.add(webElement.getText());
+        }
+
+        return items;
+    }
+
+    public void viewDeployment(String deploymentName) {
```
Comment (zebrik) [inline common/src/main/java/org/jboss/hal/testsuite/page/runtime/DeploymentPage.java:63]: Please add javadoc to this method. Purpose of this method - to open deployment detail - may not be obvious.

**Review by zebrik — COMMENTED** [review]
Thank you for updates. I have only several last comments.

**Comment by zebrik** [comment]
LGTM now. Please squash and I will merge.

**Comment by jstefl** [comment]
Thank you for review, squashed.

**Comment by zebrik** [comment]
@jstefl can you please change last commit message to something more suitable for upstream?

**Comment by jstefl** [comment]
Commit message fixed.

**Comment by zebrik** [comment]
Thank you for patience. Merging.
